lib/kafka_ex/new/cluster_metadata.ex

```diff
@@ -0,0 +1,268 @@
+defmodule KafkaEx.New.ClusterMetadata do
+  @moduledoc """
+  Encapsulates what we know about the state of a Kafka broker cluster
+
+  This module is mainly used internally in ServerKayrock, but some of its
+  functions may be useful for extracting metadata information
+  """
+
+  alias KafkaEx.New.Broker
+  alias KafkaEx.New.Topic
+  alias KafkaEx.New.KafkaExAPI
+
+  defstruct brokers: %{},
+            controller_id: nil,
+            topics: %{},
+            consumer_group_coordinators: %{}
+
+  @type t :: %__MODULE__{
+          brokers: %{KafkaExAPI.node_id() => Broker.t()},
+          controller_id: KafkaExAPI.node_id(),
+          topics: %{KafkaExAPI.topic_name() => Topic.t()},
+          consumer_group_coordinators: %{
+            KafkaExAPI.consumer_group_name() => KafkaExAPI.node_id()
+          }
+        }
+
+  @typedoc """
+  Specifies for selecting a node in `select_node/2`
+
+  Selectors:
+
+    * `KafkaExAPI.node_id()` - A specific node (verifies that we know about the node)
+    * `:controller` - The cluster's controller node id
+    * `:random` - Selects a randon node
+    * `{:topic_partition, topic_name, partition_id}` - The controller for a
+      given partition
+    * `{:consumer_group, consumer_group_name}` - the controller for a given
+      consumer group
+  """
+  @type node_selector ::
+          KafkaExAPI.node_id()
+          | :controller
+          | :random
+          | {:topic_partition, KafkaExAPI.topic_name(),
+             KafkaExAPI.partition_id()}
+          | {:consumer_group, KafkaExAPI.consumer_group_name()}
```

Comment:
I like this.

One caveat: when reading the code using functions that take a `node_selector`, it wasn't easy to understand for me that it was indeed a node selector (for example: `{:topic partition, topic, partition}`). 

Suggestion: have a module that helps build the selector. For example:

```elixir
defmodule Node do
   def for_topic_partition(topic, partition), do: {:topic_partition, topic, partition}
   def random(), do: :random
   # etc...
end

# Now we could write :
ServerKayrock.kayrock_call(server, request, Node.for_topic_partition(topic, partition))
```

Maybe it's out of the scope of what you wish to do here, though (also maybe it's overkill ^^).

Reply:
Not overkill, I like it.

Reply:
I added a `NodeSelector` module.  I ended up using a struct with a `:strategy` field in order to make it easier to make clear pattern matches.